zipkin2/src/main/java/zipkin2/SpanBytesDecoderDetector.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2015-2018 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2;
+
+import zipkin2.codec.BytesDecoder;
+import zipkin2.codec.SpanBytesDecoder;
+
+/**
+ * Detecting decoder used in transports which don't include means to identify the type of the data.
+ *
+ * <p>For example, we can identify the encoding and also the format in http via the request path and
+ * content-type. However, in Kafka it could be that folks send mixed Zipkin data without identifying
+ * its format. For example, Kafka historically has no content-type and users don't always segregate
+ * different queues by instrumentation format.
+ */
+// In TBinaryProtocol encoding, the first byte is the TType, in a range 0-16
+// .. If the first byte isn't in that range, it isn't a thrift.
+//
+// When byte(0) == '[' (91), assume it is a list of json-encoded spans
+//
+// When byte(0) == 10, assume it is a proto3-encoded span or trace ID field
+//
+// When byte(0) <= 16, assume it is a TBinaryProtocol-encoded thrift
+// .. When serializing a Span (Struct), the first byte will be the type of a field
+// .. When serializing a List[ThriftSpan], the first byte is the member type, TType.STRUCT(12)
+// .. As ThriftSpan has no STRUCT fields: so, if the first byte is TType.STRUCT(12), it is a list.
+public final class SpanBytesDecoderDetector {
```

Comment:
cc @drolando this will be the replacement for that internal class you were looking at earlier

cc also @wu-sheng

Reply:
Sure. Travel around in the whole weekend. :)